### PR TITLE
Seller Experience-Stepper: WooCommerce upgrade eligibility data store

### LINKED ISSUE #62914: Seller Experience-Stepper: WooCommerce upgrade eligibility data store
### What

Add a data store for the WooCommerce upgrade eligibility requirements data.

### Why

This will be used by the  WooCommerce confirm step in #62714

### How

_No response_

### PR DESCRIPTION
**Pull request by markbiek**
#### Changes proposed in this Pull Request

Data store for checking WooCommerce upgrade eligibility
    
Based on `client/state/data-layer/wpcom/sites/automated-transfer/eligibility`
    
Includes selectors for:
    
- The complete set of transfer eligibility data
- Eligibility holds
- Eligibility warnings
- General subdomain warnings
- WordPress.com subdomain warnings

#### Testing instructions

`yarn test-packages packages/data-stores/src/woocommerce-eligibility/test/`

Related to #62714
Fixes #62914

### CODE REVIEW

File: packages/data-stores/src/woocommerce-eligibility/reducer.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import { combineReducers } from '@wordpress/data';
+import { TransferEligibility } from './types';
+import type { Action } from './actions';
+import type { Reducer } from 'redux';
+
+export const transferEligibility: Reducer<
+	{ [ key: number ]: TransferEligibility } | undefined,
+	Action
+> = ( state, action ) => {
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/reducer.ts]: A more elegant way to type the reducer is:
```ts
export type State = {
  [ key: number ]: TransferEligibility
};

export const transferEligibility: Reducer< State, Action > = ( state = {}, action ) => {
  ...
}
```
- the `State` type can be defined directly, instead of being calculated with `ReturnType`, `typeof` etc.
- it doesn't need to include the `| undefined` bit, because the `Reducer` type adds that automatically to its type parameter. Also, `State | undefined` is only the reducer input, the return type is always `State`, without the `undefined`.
- to avoid ever returning an `undefined` type, the `state` parameter needs to have a default. In this case, it's `{}`.

File: packages/data-stores/src/woocommerce-eligibility/reducer.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+				[ action.siteId ]: action.transferEligibility,
+			};
+		}
+	}
+
+	return state;
+};
+
+// Tracks product list fetching state
+export const isFetchingTransferEligibility: Reducer< boolean | undefined, Action > = (
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/reducer.ts]: The type is `Reducer< boolean >`. The reducer state is never `undefined`. Except for a short time during initialization, but that's invisible to the user.

File: packages/data-stores/src/woocommerce-eligibility/resolvers.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import { wpcomRequest } from '../wpcom-request-controls';
+import { STORE_KEY } from './constants';
+import { TransferEligibility } from './types';
+
+/**
+ * Requests the site's transfer eligibility from the WPCOM API.
+ *
+ * @param   {number} siteId The id of the site to retrieve
+ */
+export function* getAutomatedTransferEligibility( siteId: number ) {
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: Generators and controls are deprecated in latest `@wordpress/data`. Let's use thunks here:
```js
import wpcomRequest from 'wpcom-proxy-request';

const getAutomatedTransferEligibility = ( siteId ) => async ( { dispatch } ) => {
  dispatch.requestTransferEligibility( siteId );
  const transferEligibility = await wpcomRequest( { path, apiVersion } );
  dispatch.receiveTransferEligibility( transferEligibility, siteId );
}
```

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+export const getState = ( state: State ) => state;
+
+export const getAutomatedTransferEligibility = (
+	state: State,
+	siteId: number | null
+): TransferEligibility | null => {
+	if ( ! state || ! state.transferEligibility ) {
+		return null;
+	}
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: `state.transferEligibility` always exists and always is an object. That's guaranteed by the store and doesn't need to be checked.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	return state.transferEligibility[ siteId ];
+};
+
+export const getEligibilityHolds = (
+	state: State,
+	siteId: number | null
+): TransferEligibilityError[] | null => {
+	const transferEligibility = select( STORE_KEY ).getAutomatedTransferEligibility( siteId );
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Selectors are supposed to read only from `state`, never from anywhere else:
```js
const transferEligibility = getAutomatedTransferEligibility( state, siteId );
```

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		return null;
+	}
+
+	const { errors } = transferEligibility;
+
+	if ( ! errors ) {
+		return null;
+	}
+
+	errors.map( ( { code }: { code: string } ) => {
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: The `TransferEligibilityError` already declares what the type of `code` is and TypeScript knows that. You can write just:
```ts
errors.map( ( { code } ) => ... );
```
Alos, the result of the `.map` is not used, so why is it called? Maybe you want to return the result instead of the original `errors`?

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+	if ( ! warnings ) {
+		return null;
+	}
+
+	// combine plugin and theme warnings into one list
+	let combined: TransferEligibilityWarning[] = [];
+	for ( const key of Object.keys( warnings ) ) {
+		combined = combined.concat( warnings[ key ] );
+	}
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Can be written as:
```js
const combined = Object.values( warnings ).flat();
```

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	state: State,
+	siteId: number | null
+): TransferEligibilityWarning[] | null => {
+	const eligibilityWarnings = select( STORE_KEY ).getEligibilityWarnings( siteId );
+
+	if ( ! eligibilityWarnings ) {
+		return null;
+	}
+
+	return eligibilityWarnings?.filter( ( { id } ) => id !== 'wordpress_subdomain' ) ?? null;
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: The `?.` and `??` operators seem to be redundant because an `undefined` or `null` value has already been handled few lines above.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	state: State,
+	siteId: number | null
+): TransferEligibilityWarning | null => {
+	const eligibilityWarnings = select( STORE_KEY ).getEligibilityWarnings( siteId );
+
+	if ( ! eligibilityWarnings ) {
+		return null;
+	}
+
+	const warning = eligibilityWarnings?.find( ( { id } ) => id === 'wordpress_subdomain' );
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Same here, `?.` is redundant.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
+import { select } from '@wordpress/data';
+import { get } from 'lodash';
+import { STORE_KEY, statusMapping } from './constants';
+import { TransferEligibilityWarning, TransferEligibilityError, TransferEligibility } from './types';
+import type { State } from './reducer';
+
+export const getState = ( state: State ) => state;
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Such a selector shouldn't be needed anywhere.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	const { errors } = transferEligibility;
+
+	if ( ! errors ) {
+		return null;
+	}
+
+	errors.map( ( { code }: { code: string } ) => {
+		return get( statusMapping, code, '' );
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Let's avoid use of Lodash and use `statusMapping[ code ] ?? ''` instead.

File: packages/data-stores/src/woocommerce-eligibility/reducer.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	state = false,
+	action
+) => {
+	switch ( action.type ) {
+		case 'TRANSFER_ELIGIBILITY_REQUEST':
+			return true;
```
Comment (jsnajdr) [inline packages/data-stores/src/woocommerce-eligibility/reducer.ts]: The "is fetching" state should be tracked by `siteId`. We can issue many parallel requests for different `siteId`'s that are completely independent.

Also, you can avoid it completely because `@wordpress/data` has built-in selectors like `isResolving` that track this automatically.

File: packages/data-stores/src/index.ts [hunk 13]
```diff
@@ -12,6 +12,7 @@ import * as Site from './site';
 import * as User from './user';
 import * as Verticals from './verticals';
 import * as VerticalsTemplates from './verticals-templates';
+import * as AutomatedTransferEligibility from './woocommerce-eligibility';
```
Comment (jsnajdr) [inline packages/data-stores/src/index.ts]: The name of the folder should be `atomic-eligibility` or `automated-transfer-eligibility` because that's what it does. There is noting Woo-specific there.

**Review by jsnajdr — CHANGES_REQUESTED** [review]

File: packages/data-stores/src/index.ts (hunk 13 above)
Comment (tyxla) [reply packages/data-stores/src/index.ts]: I second this, and there are a few occurrences in the code itself as well (e.g. store key for example)

File: packages/data-stores/src/woocommerce-eligibility/constants.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+export const STORE_KEY = 'automattic/woocommerce-eligibility';
+
+export const eligibilityHolds = {
```
Comment (tyxla) [inline packages/data-stores/src/woocommerce-eligibility/constants.ts]: Do we really need to export this one? I'd always encourage not exporting something when it's not necessary to be consumed from outside.

File: packages/data-stores/src/woocommerce-eligibility/constants.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Maps the constants used in the WordPress.com API with
+ * those used inside of Calypso. Somewhat redundant, this
+ * provides safety for when the API changes. We need not
+ * changes the constants in the Calypso side, only here
+ * in the code directly dealing with the API.
+ */
+export const statusMapping = {
+	blocked_atomic_transfer: eligibilityHolds.BLOCKED_ATOMIC_TRANSFER,
```
Comment (tyxla) [inline packages/data-stores/src/woocommerce-eligibility/constants.ts]: Seems like there's not a good reason to hold the eligibility reasons in a `eligibilityHolds` object. They can be constants, accessible only to the module, or even better - inlined if there's no other use for them when declared separately as constants.

File: packages/data-stores/src/woocommerce-eligibility/test/actions.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+describe( 'actions', () => {
+	it( 'should return a TRANSFER_ELIGIBILITY_REQUEST action', () => {
+		const expected = {
+			type: 'TRANSFER_ELIGIBILITY_REQUEST',
+		};
+
+		expect( requestTransferEligibility() ).toEqual( expected );
+	} );
+
+	it( 'should return a TRANSFER_ELIGIBILITY_RECEIVE action', () => {
```
Comment (tyxla) [inline packages/data-stores/src/woocommerce-eligibility/test/actions.ts]: Noting that tests that only test if an action creator function returns the same object it returns, are generally redundant, and seem to add more noise than actually help ensure functionality works.

File: packages/data-stores/src/woocommerce-eligibility/test/reducer.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import { transferEligibility } from '../reducer';
+import { TransferEligibility } from '../types';
+
+describe( 'reducer', () => {
+	it( 'returns the correct default state', () => {
+		const state = transferEligibility( undefined, { type: 'TEST_ACTION' } );
+		expect( state ).toEqual( undefined );
+	} );
+
+	it( 'returns the site transfer elibibility', () => {
```
Comment (tyxla) [inline packages/data-stores/src/woocommerce-eligibility/test/reducer.ts]: Typo here:

```suggestion
	it( 'returns the site transfer eligibility', () => {
```

Also, I'd rephrase this to something like:

```suggestion
	it( 'keys transfer eligibility by site ID', () => {
```

File: packages/data-stores/src/woocommerce-eligibility/test/reducer.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+		};
+
+		const state = transferEligibility( undefined, {
+			type: 'TRANSFER_ELIGIBILITY_RECEIVE',
+			siteId,
+			transferEligibility: expectedTransferEligibility,
+		} );
+
+		expect( state ).toEqual( {
+			[ siteId ]: expectedTransferEligibility,
```
Comment (tyxla) [inline packages/data-stores/src/woocommerce-eligibility/test/reducer.ts]: When testing such reducers, in addition to a test that ensures that keying by site ID works, we could also benefit from testing those scenarios:

* Data is updated properly when passed for a site ID that is already in the state.
* Data is added properly for a new site, while data for existing site IDs is kept intact.

File: packages/data-stores/src/woocommerce-eligibility/resolvers.ts (hunk 3 above)
Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts:11]: @alshakero All of the existing data stores are using generators.

Do you have any thoughts on switching to thunks?

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts (hunk 5 above)
Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: I may have misunderstood but I thought it was necessary to do `select( STORE_KEY ).getAutomatedTransferEligibility( siteId )` to ensure that the corresponding resolver was called in the event the state hadn't been populated yet.

@alshakero Can you let me know if I'm mistaken about that?

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts (hunk 6 above)
Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: This was code ported over from elsewhere but I can clean that bit up.

File: packages/data-stores/src/index.ts (hunk 13 above)
Comment (markbiek) [reply packages/data-stores/src/index.ts]: That makes sense. I can adjust the naming.

File: packages/data-stores/src/woocommerce-eligibility/constants.ts (hunk 15 above)
Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/constants.ts]: This was code ported over from elsewhere but I can try and clean it up a bit.

File: packages/data-stores/src/woocommerce-eligibility/resolvers.ts (hunk 3 above)
Comment (alshakero) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: Quoting slack response: 
> Wow, I had no idea they were deprecated. For new stores, I’d go for thunks for sure then. And maybe we need a project to upgrade the stores away from generators.

Comment (jsnajdr) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: Like, they are not formally deprecated in a "wrapped with `@wordpress/deprecated` and triggering a warning" sense, but Gutenberg itself has been fully migrated away and we should migrate plugins, too, or at least, don't introduce new usages.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts (hunk 5 above)
Comment (alshakero) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: I think you're right. Maybe @jsnajdr didn't know this selector has a resolver?

Calling `getAutomatedTransferEligibility` directly won't trigger the resolver because the store has no way to know you called `getAutomatedTransferEligibility` to go ahead and trigger the resolver.

But when you do `select( STORE_KEY ).getAutomatedTransferEligibility( siteId )` you go through the store, giving it a chance to do whatever it needs to do (e.g., trigger the resolver).

I think this is one of the two ways to chain selectors, this and [createRegistrySelector](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#createregistryselector).

Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Splendid, thanks for clarifying!

File: packages/data-stores/src/woocommerce-eligibility/resolvers.ts (hunk 3 above)
Comment (markbiek) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: @jsnajdr Yeah, if thunks are the direction we're going, we should definitely make sure new code uses them.

Can you offer any advice on how to handle typing that `dispatch` object? I'm struggling a bit with that.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts (hunk 5 above)
Comment (jsnajdr) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Oh, I see, you need to trigger the resolver. Then you can either use `createRegistrySelector`, or create a "forwarding" resolver for every selector that depends on the main one:
```js
const derivedSelector = ( state ) => originalSelector( state ) * 2;
const derivedResolver = ( /* args */ ) => ( { resolveSelect } ) => resolveSelect.originalSelector( /* args */ );
```

The trouble with global `select( STORE_KEY )` is that it always refers to the instance registered in the default registry. But you can create many registries where each has its own instance and in that case these selectors wouldn't work. It's like not using `this.foo()` on an object instance and using a global shared variable instead.

File: packages/data-stores/src/automated-transfer-eligibility/reducer.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+			};
+		}
+	}
+
+	return state;
+};
+
+const reducer = combineReducers( {
+	transferEligibility,
+} );
```
Comment (jsnajdr) [inline packages/data-stores/src/automated-transfer-eligibility/reducer.ts]: Now when there's only one sub-reducer, you don't need to combine it. Just make `transferEligibility` itself into the store reducer.

File: packages/data-stores/src/automated-transfer-eligibility/index.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+export type { State };
+export { STORE_KEY };
+
+let isRegistered = false;
+export function register(): typeof STORE_KEY {
+	if ( ! isRegistered ) {
+		isRegistered = true;
+		registerStore< State >( STORE_KEY, {
+			actions,
+			controls: controls as any,
```
Comment (jsnajdr) [inline packages/data-stores/src/automated-transfer-eligibility/index.ts]: After migration of the resolver from controls to thunks, the `controls` registration can be removed, too.

File: packages/data-stores/src/woocommerce-eligibility/selectors.ts (hunk 5 above)
Comment (alshakero) [reply packages/data-stores/src/woocommerce-eligibility/selectors.ts]: Oh great explanation. This is even more relevant in data-stores package because it's meant to be as reusable as possible. I'd go with `createRegistrySelector`. It's clean and minimalistic.

/cc @razvanpapadopol I remember calling global `select` was your solution, you might want to know this.

File: packages/data-stores/src/automated-transfer-eligibility/reducer.ts (hunk 19 above)
Comment (andres-blanco) [reply packages/data-stores/src/automated-transfer-eligibility/reducer.ts]: I've been having trouble typing the State without the `combineReducers` call. I am working to find a solution

Comment (alshakero) [reply packages/data-stores/src/automated-transfer-eligibility/reducer.ts]: `type State = ReturnType< typeof transferEligibility >` does this work?

Comment (andres-blanco) [reply packages/data-stores/src/automated-transfer-eligibility/reducer.ts]: No, that didn't work. What was wrong here was trying to use the state as if it was combined in the reducer. I just removed the `RootState` type and everything worked :). Ref: https://github.com/Automattic/wp-calypso/pull/62861/commits/f6456d579b3f17b0aca1907cd0930b2a035ff1ca

**Review by jsnajdr — COMMENTED** [review]
The global `select( STORE_KEY )` in selectors is still a problem that should be fixed. Other than that, the PR is in a shippable state. Thanks for addressing the review comments!

File: packages/data-stores/src/woocommerce-eligibility/resolvers.ts (hunk 3 above)
Comment (andres-blanco) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: It seems that Calypso is running version `6.1.5` of the `@wordpress/data` package. In this versions thunks don't seem to be enabled. I will be doing some more testing but it feels like we wont be able to go this way unless we update the `@wordpress/data` package.

Comment (jsnajdr) [reply packages/data-stores/src/woocommerce-eligibility/resolvers.ts]: In that version, you need to enable them with a flag:
```js
registerStore( name, {
  reducer,
  actions,
  selectors,
  __experimentalUseThunks: true
} );
```
Only the latest version of `@wordpress/data`, shipped in WordPress 6.0, enables them by default.

**Comment by andres-blanco** [comment]
> The global `select( STORE_KEY )` in selectors is still a problem that should be fixed. Other than that, the PR is in a shippable state. Thanks for addressing the review comments!

@jsnajdr Thanks for all the input so far! I've replaced the global `select` calls by using `createRegistrySelector` but I've lost all type aid because that function doesn't seem to be correctly typed: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__data/index.d.ts#L171-L174. I think that type should be replaced by something like this:
```js
export function createRegistrySelector<S extends typeof select, T, U extends (state: any, ...args: any[]) => T >(
    registrySelector: (select: S) => U,
): U;
````
But I don't know what's the process to submit an update to that repository.

**Comment by jsnajdr** [comment]
> I've lost all type aid because that function doesn't seem to be correctly typed

We're working on adding types directly to the `@wordpress/data` package and deprecating the separate `@types` one. We'll eventually add the correct type for `createRegistrySelector`, too.

**Review by jsnajdr — APPROVED** [review]
:+1:

**Comment by andres-blanco** [comment]
> > I've lost all type aid because that function doesn't seem to be correctly typed
> 
> We're working on adding types directly to the `@wordpress/data` package and deprecating the separate `@types` one. We'll eventually add the correct type for `createRegistrySelector`, too.

That sounds like a great idea. Thanks for everything :)
